### DISCUSSION TITLE
How is rtcontrol aware of fields?

### DISCUSSION THREAD

**Discussion by Rouneq** [kannibalox/pyrosimple#66] [Q&A] [answered]
Or to put it another way, can I use non-aware fields with rtcontrol? Let's say I want to cull out some torrents I've been seeing a while that have plenty of other seeders. From what I can tell, the number of seeders/leechers isn't a piece of info I normally have access to. Can I define new fields for rtcontrol awareness or am I stuck?

FWIW, I know I can write a script. But the main question is of interest to me.

**Comment by kannibalox** [comment]
Yep, you can use custom fields to do that, this page has an example of what you're probably looking for with the `seeds`/`leechers` fields: https://kannibalox.github.io/pyrosimple/custom-code/#custom-fields. I've considered making it a core field, but I've been a little concerned that it might be misinterpreted as the number of actively connected peers, which would be a different call.

All of the core fields are actually defined in a similar manner, if that helps to understand things: https://github.com/kannibalox/pyrosimple/blob/main/src/pyrosimple/torrent/engine.py#L333

**Reply by Rouneq** [reply]
You know, I looked at the Advanced page and saw "custom" and my mind when to "That's about `d.custom`" and moved on. Thanks for the redirection.

Alright. I know how to go about defining my own fields. Now I'm trying to understand where the data I'm after actually lives. So far, this is what I'm seeing.

```
rtxmlrpc -o repr d.multicall.filtered '' default 'equal=d.hash=,cat=2CC51D61B18D22254B5446753AC65000EDADF58D' d.name= 
d.connection_seed= d.disconnect.seeders= d.down.choke_heuristics.seed= d.up.choke_heuristics.seed= d.connection_leech= 
d.up.choke_heuristics.leech= d.peers_accounted= d.peers_complete= d.peers_connected= d.peers_not_connected=
[['The.Awesome.Torrent.Data',
  '',
  0,
  '',
  '',
  '',
  '',
  0,
  0,
  0,
  0]]
```

Where rutorrent reports

```
| Name                     | Seeds   | Peers |
| The.Awesome.Torrent.Data | 0 (111) | 0 (0) |
```

I suspect I'm missing something important here. rutorrent appears to be referencing the `peer` commands for this data, but I threw in the others just to get a feel for what's there.

**Reply by kannibalox** [reply, marked as answer]
This is a simple `config.py` that would get you what you're looking for:
```py
from pyrosimple.torrent import engine
import pyrosimple


def _custom_fields():
    from pyrosimple.torrent import engine
    from pyrosimple.util import fmt, matching
    import functools

    # Add rTorrent attributes not available by default                                                                                  
    def get_tracker_field(obj, name, aggregator=sum):
        "Get an aggregated tracker field."
        return aggregator(obj.rpc_call("t.multicall", ["", f"t.{name}="])[0])

    yield engine.DynamicField(
        int,
        "downloaders",
        "number of completed downloads",
        matcher=matching.FloatFilter,
        accessor=lambda o: get_tracker_field(o, "scrape_downloaded"),
        requires=["t.multicall=,t.scrape_downloaded="],
    )
    yield engine.DynamicField(
        int,
        "seeds",
        "number of seeds",
        matcher=matching.FloatFilter,
        accessor=lambda o: get_tracker_field(o, "scrape_complete"),
        requires=["t.multicall=,t.scrape_complete="],
    )
    yield engine.DynamicField(
        int,
        "leeches",
        "number of leeches",
        matcher=matching.FloatFilter,
        accessor=lambda o: get_tracker_field(o, "scrape_incomplete"),
        requires=["t.multicall=,t.scrape_incomplete="],
    )

# Register our custom fields to the proxy                                                                                               
for field in _custom_fields():
    engine.TorrentProxy.add_field(field)
```
Then you could find all torrents with more than 10 seeders:
```
rtcontrol seeds\>10
```

There's some python magic going on to simplify the code, but essentially e.g. the `seeds` field is the sum of the result of the sum of `t.scrape_complete` for all the trackers. If you wanted to see that directly from rtxmlrpc:
```sh
rtxmlrpc d.multicall2 '' main d.hash= t.multicall=,t.scrape_complete=
```

The `t.scrape_*` commands rely on information being scraped directly from the tracker, which is why this section of the config is pretty important for those stats to be up to date: https://github.com/kannibalox/pyrosimple/blob/main/src/pyrosimple/data/full-example.rc#L51 (by default rtorrent just does a single scrape at start-up). In contrast, the `d.peers_*` commands only report stats for the *actively* connected peers, which can still be useful (and you could make a custom field for it), but probably not what you're looking for in this case.

ruTorrent is actually using both stats, so if you had `1 (27)` in the seeds column, that would mean the tracker reports there are 27 seeds, but only 1 is actually connected to you.

**Reply by Rouneq** [reply]
Wow. Thanks. I wasn't expecting a full config loadout, but I'm glad you provided it. There were more dots to connect than I knew about.

**Reply by kannibalox** [reply]
Also, just a heads up: `d.disconnect.seeders` is an action, not a stat, it immediately disconnects any connected seeders.
